cycle.1
.\" .ig
.\" cycle manual page.
.\" 
.\" Miriam Ruiz <little_miry@yahoo.es>, 2005.
.\" 
.\" This next line loads the www macro package so that the .URL macro can be used.
.\" ..
.\" .mso www.tmac
.TH "cycle" "1" "6 April 2005" "Miriam Ruiz" "Utils"
.SH "NAME"
.
cycle \- a calendar for women
.
.SH "SYNOPSIS"
.
.B cycle
.
.SH "DESCRIPTION"
.
Cycle is a calendar program for women. Given a cycle length or statistics for several periods, it can calculate the days until menstruation, the fertile period, and the days to ovulations, and define the d.o.b. (date of birth) of a child. It allows the user to write notes and helps to supervise the reception of hormonal contraceptive tablets.
.
.SH "FEATURES"
The program is capable of:
.P
 \- Calculate of the days of menstruation, based on length of the cycle or on statistics of a period.
 \- Calculate the fertile period and day to ovulations
 \- Definition of D.O.B . (Date Of Birth) of a child
 \- Allows to write notes
 \- Helps to supervise reception of hormonal contraceptive tablets
.P
.SH "HOW DOES IT WORK"
It uses the calendar method for the definition of fertile days. For this is necessary to determine duration, at least, of last six cycles. For determine the fertile days it uses this algoritm:
.P
 \- First day: duration of shortest cycle minus 18
 \- Last day: duration of longest cycle minus 11
 \- Ovulation: is considered in the middle of fertile period (Determine the exact day isn't possible with the calendar method) 
.P

Date of a birth of the child is counted from the beginning of last menstruation (in obstetrics are accepted 40 weeks). This date is corrected with the duration of a cycle (are subtracted or added same number of days that a cycle differs from 28 days). Is possible check the results at: http://cir.msk.ru/sroki.shtml (in Russian)
.SH "USAGE"
.
It is necessary to mark the beginning of menstruation. This is possible clicking with the right button on the date and chosing "The beginning of a cycle". Similarly the mark can be removed. Using duration of last six cycles, the program will calculate

.P
 \- Fertile period \- green cells
 \- Ovulation day (the middle fertile period) \- brightly green cell
 \- Beginning of following menstruation \- pink cells 
.P

Cycle duration is set in options or is calculated an average of last six periods.

To determine prospective date of a birth of the child, it is necessary first day of last cycle to mark in addition as "Pregnancy".

Depending on your purposes, you can display fertile days. Information about interesting day can be received clicking with left button.

Is possible adding a note to any date using corresponding item of the menu. Day with a note are underlined.

It is necessary to note that factor of failures of calendar contraception method is about 10 percent. This method is suitable only for women with enough regular cycle.

If your doctor has recommended hormonal contraceptive this program will help to supervise their reception. Closely familiarize with the instruction applied to the chosen preparation. Packing can contain 21 tablet (the circuit of reception 21 day on one tablet, then 7 days pause) or 28 tablets in this case all tablets are accepted without interruption every day of 28. In the program, is marked only day of reception of each first tablet of packing.

NOTE: This program is not a reliable contraceptive method. It does neither help to prevent sexual transmision diseases like AIDS. It is just an electronic means of keeping track of some of your medical data and extract some statistical conclusions from them. You cannot consider this program as a substitute for your gynecologist in any way.
.
.SH "FILES"
.
.TP 
.I ~/.cycle
This is the directory where cycle stores the data.
.
.SH "AUTHOR"
.
cycle was written by Oleg S. Gints and is currently maintained by Matt Molyneaux.
.P
This manual page was written for Debian by Miriam Ruiz <little_miry@yahoo.es>.
.
.SH "SEE ALSO"
.
You can find cycle's home page at: 
.URL http://moggers.co.uk/cgit/cycle.git/about
http://moggers.co.uk/cgit/cycle.git/about
.P
More information about calendar method are available at:
.URL http://www.mama.ru/gynecolog/STA/st18.htm
http://www.mama.ru/gynecolog/STA/st18.htm (in Russian)
.


CHANGELOG
#====================================================
#	Cycle - calendar for women
#	Distributed under GNU Public License
# Original author: Oleg S. Gints
# Maintainer: Matt Molyneaux (moggers87+git@moggers87.co.uk)
# Home page: http://moggers.co.uk/cgit/cycle.git/about
#===================================================
0.3.2 (19.09.2011)
    - fixed issues with mixing of tabs and spaces, all code now uses 4 spaces as
      a tab
    - merged patches found in Fedora RPM (originated from Debian)
    - removed Russian and German readme files, I don't speak either well enough
    - added some very basic printing
    - moved to a different website as original author hasn't replied to emails
      and SourceForge don't deal with abandoned projects anymore
0.3.1 (15.09.2005)
    - change to Python 2.4 - repleace rotor module
    - restore old default color scheme
    - The description of the program in English from
      Marco Papa Manzillo<mpapamanz@users.sourceforge.net> is corrected
    - Now the name of the user is stored in a file with data.
    
    from  Miriam Ruiz <little_miry@yahoo.es>:
    - fixed colors not initialized the first time a user was created.
    - Added code to use wx 2.5.3 or higher if multiple versions are installed.
    - Added es, ca and gl translations, and solved some bugs in locale subsystem.
    - Added note alerting that this is not a reliable contraceptive method every
      time a new user is created.
    - Added man page
    - fixed Cycle for unicode version of wx

0.3.0 (21.12.2004)
    - Add translation to the german language
      from Christian Weiske <cweiske@users.sourceforge.net>
    - change to wxPython = 2.5.X

0.2.1 (06.06.2004)
    - Add colours settings

0.2.0 (28.04.2004)
    - Add translation for czech and slovak language
      from Jozef Riha <zefo@seznam.cz>
    - change to Python 2.3
    - Add description for english language
      from Marco Papa Manzillo <mpapamanz@users.sourceforge.net>

0.0.5 (30.12.2002)
    - fix PopupMenu for new wxPythonGTK-2.3.3.1
    - fix exception from LANGUAGE=lang1:lang2

0.0.4 (24.10.2002)
    - add setup.py and MANIFEST.in, now make new release so:
      python setup.py sdist
    - add check reception of hormonal contraceptive tablets

0.0.3 (16.05.2002)
    - Add icons
    - Convert README* in *.html

0.0.2 (05.05.2002)
    - Add line under toolbar
    - Change color legend "today"
    - Change BackgroundColour = LIGHT BLUE
    - clean up code
    - Add Note
    - Add calc birthday
    - encode messages to locale
    - Add method "by average"
    - Add cycle.spec

0.0.1 (22.04.2002)
    - First release
    
    


src/cycle.py
#====================================================
#        Cycle - calendar for women
#        Distributed under GNU Public License
# Original author: Oleg S. Gints
# Maintainer: Matt Molyneaux (moggers87+git@moggers87.co.uk)
# Home page: http://moggers.co.uk/cgit/cycle.git/about
#===================================================    
import os, sys, gettext
import locale

import wxversion 
wxversion.ensureMinimal('2.5.3')
import wx
import wx.html
import wx.lib.colourdb

from cal_year import *
from save_load import *
from dialogs import *
from set_dir import *

import gettext
import __builtin__
lang_find = False
if not '__WXMSW__' in wx.PlatformInfo:
    for lang_env_var in ('LANGUAGE', 'LC_ALL', 'LC_CTYPE', 'LANG'):
        if lang_find:
            break
        if os.environ.has_key(lang_env_var):
            env_language = os.environ[lang_env_var]
            for s_lang in env_language.split(':'): # if set more languages
                os.environ[lang_env_var] = s_lang
                try:
                    dl = locale.getdefaultlocale()
                    lang = [dl[0][0:2]]
                    l = gettext.translation('cycle', msg_dir, lang)
                    if wx.USE_UNICODE:
                        __builtin__.__dict__['_'] = lambda s: l.ugettext(s)
                    else:
                        __builtin__.__dict__['_'] = lambda s: l.ugettext(s).encode(dl[1])
                    _('try decode this string')
                    lang_find = True
                    break #language was found
                except:
                    pass
else: #for MS Windows
    try:
        dl = locale.getdefaultlocale()
        lang = [ dl[0][0:2] ]
        l = gettext.translation('cycle', msg_dir, lang)
        if wx.USE_UNICODE:
            __builtin__.__dict__['_'] = lambda s: l.ugettext(s)
        else:
            __builtin__.__dict__['_'] = lambda s: l.ugettext(s).encode(dl[1])
        _('try decode this string')
        lang_find = True
    except:
        pass

if not lang_find:
    __builtin__.__dict__['_'] = lambda s: s
    lang = [""]


class MyFrame(wx.Frame):
    def __init__(self, parent, ID, title):
        wx.Frame.__init__(self, parent, ID, title,
                       wx.DefaultPosition, wx.Size(800, 600))

        wx.Image_AddHandler(wx.PNGHandler())
        self.printer = wx.html.HtmlEasyPrinting()
        icon = wx.Icon(os.path.join(icons_dir, 'mini/cycle.xpm'), wx.BITMAP_TYPE_XPM)
        self.SetIcon(icon)

        Val.frame = self
        self.CreateStatusBar()
        self.MakeToolMenu()  # toolbar
        
        self.cal = Cal_Year(self)
        self.OnCurrent(self)
        wx.EVT_CLOSE(self, self.OnCloseWindow)

    def OnCloseWindow(self, event):
        Save_Cycle(cycle.name, cycle.passwd, cycle.file)
        self.Destroy()

    def TimeToQuit(self, event):
        self.Close(True)

    def MakeToolMenu(self):
        tb = self.CreateToolBar(wx.TB_HORIZONTAL | wx.NO_BORDER)
        tb.SetToolBitmapSize(wx.Size(24, 24))

        SetToolPath(self, tb, 10, os.path.join(bitmaps_dir, 'dec.png'), _('Dec Year'))
        wx.EVT_TOOL(self, 10, self.OnDecYear)

        SetToolPath(self, tb, 20, os.path.join(bitmaps_dir, 'curr.png'), _('Current Year'))
        wx.EVT_TOOL(self, 20, self.OnCurrent)

        SetToolPath(self, tb, 30, os.path.join(bitmaps_dir, 'inc.png'), _('Inc Year'))
        wx.EVT_TOOL(self, 30, self.OnIncYear)

        tb.SetToolSeparation(50)
        tb.AddSeparator()

        SetToolPath(self, tb, 40, os.path.join(bitmaps_dir, 'legend.png'), _('Legend'))
        wx.EVT_TOOL(self, 40, self.Legend)
        
        SetToolPath(self, tb, 45, os.path.join(bitmaps_dir, 'export.png'), _('Export to iCal'))
        wx.EVT_TOOL(self, 45, self.Export)
        
        SetToolPath(self, tb, 50, os.path.join(bitmaps_dir, 'set.png'), _('Settings'))
        wx.EVT_TOOL(self, 50, self.Settings)
        
        SetToolPath(self, tb, 55, os.path.join(bitmaps_dir, 'help.png'), _('Help'))
        wx.EVT_TOOL(self, 55, self.Info)

        SetToolPath(self, tb, 57, os.path.join(bitmaps_dir, 'print.png'), _('Print'))
        wx.EVT_TOOL(self, 57, self.test)

        tb.AddSeparator()

        SetToolPath(self, tb, 60, os.path.join(bitmaps_dir, 'exit.png'), _('Exit'))
        wx.EVT_TOOL(self, 60, self.TimeToQuit)

        tb.Realize()

    def test(self, event):
        rpt = report_year(self.cal.year)
        self.printer.PreviewText(rpt)
        #dlg = Colours_Dlg(self)
        #dlg.ShowModal()
        #dlg.Destroy()

    def Legend(self, event):
        dlg = Legend_Dlg(self)
        dlg.ShowModal()
        dlg.Destroy()

    def Export(self, event):
        dlg = wx.FileDialog(self, _("Export to iCal"),
                            style = wx.SAVE)

        if dlg.ShowModal() == wx.ID_OK:
            try:
                fileobj = file(dlg.GetPath(), "w")
                report_year_ical(self.cal.year, fileobj)
            except (IOError, OSError), err:
                wx.MessageDialog(
                    self, str(err), _("Unable to export"), style=wx.OK).ShowModal()
        
    def Settings(self, event):
        dlg = Settings_Dlg(self)
        if dlg.ShowModal() == wx.ID_OK:
            self.cal.Set_Year(wx.DateTime_Today().GetYear())
        dlg.Destroy()

    def Info(self, event):
        global lang
        f_name = os.path.join(doc_dir, "README_" + lang[0] + ".html")
        if not os.path.isfile(f_name):
            f_name = os.path.join(doc_dir, "README.html")
        f = open(f_name, "r")
        msg = f.read()
        dlg = Help_Dlg(self, _('Help'), msg)
        dlg.ShowModal()


    # increment and decrement toolbar controls
    def OnIncYear(self, event):
        self.cal.Inc_Year()

    def OnDecYear(self, event):
        self.cal.Dec_Year()

    def OnCurrent(self, event):
        self.cal.Set_Year(wx.DateTime_Today().GetYear())



#----------------------------------------------
def SetToolPath(self, tb, id, bmp, title):
    global dir_path
    tb.AddSimpleTool(id, wx.Bitmap(os.path.join(dir_path, bmp), wx.BITMAP_TYPE_PNG),
                     title, title)


class MyApp(wx.App):
    def OnInit(self):
        wx.lib.colourdb.updateColourDB()
        ret = first_login()
        if ret == 'bad_login':
            return True
        elif ret == 'not_first':
            dlg = Login_Dlg(None)
            if dlg.ShowModal() == wx.ID_CANCEL:
                dlg.Destroy()
                return True
            dlg.Destroy()
        self.frame_init()
        return True

    def frame_init(self):
        frame = MyFrame(None, -1, "")
        frame.Show(True)
        self.SetTopWindow(frame)

def main():
    """Entry point"""
    global dir_path
    dir_path = os.getcwd()
    locale.setlocale(locale.LC_ALL, "")
    app = MyApp(0)
    app.MainLoop()
    return 0

if __name__ == '__main__':
    main()
